### PR TITLE
config: Handle rich (non-string) types in TOML file

### LINKED ISSUE #3538: Crash when pyproject.toml disable = [ "array", "of", "items" ]
### Steps to reproduce
1. Specify an array of messages to disable in `pyproject.toml`
2. Run pylint

### Current behavior
Crash parsing `pyproject.toml`:

```bash
$ pylint sample
Traceback (most recent call last):
  File "/Users/bruce/Library/Caches/pypoetry/virtualenvs/ifb-partner-api-AO4b-1bH-py3.8/bin/pylint", line 10, in <module>
    sys.exit(run_pylint())
  File "/Users/bruce/Library/Caches/pypoetry/virtualenvs/ifb-partner-api-AO4b-1bH-py3.8/lib/python3.8/site-packages/pylint/__init__.py", line 22, in run_pylint
    PylintRun(sys.argv[1:])
  File "/Users/bruce/Library/Caches/pypoetry/virtualenvs/ifb-partner-api-AO4b-1bH-py3.8/lib/python3.8/site-packages/pylint/lint/run.py", line 300, in __init__
    linter.load_config_file()
  File "/Users/bruce/Library/Caches/pypoetry/virtualenvs/ifb-partner-api-AO4b-1bH-py3.8/lib/python3.8/site-packages/pylint/config.py", line 785, in load_config_file
    for option, value in parser.items(section):
  File "/usr/local/Cellar/python@3.8/3.8.2/Frameworks/Python.framework/Versions/3.8/lib/python3.8/configparser.py", line 859, in items
    return [(option, value_getter(option)) for option in orig_keys]
  File "/usr/local/Cellar/python@3.8/3.8.2/Frameworks/Python.framework/Versions/3.8/lib/python3.8/configparser.py", line 859, in <listcomp>
    return [(option, value_getter(option)) for option in orig_keys]
  File "/usr/local/Cellar/python@3.8/3.8.2/Frameworks/Python.framework/Versions/3.8/lib/python3.8/configparser.py", line 855, in <lambda>
    value_getter = lambda option: self._interpolation.before_get(self,
  File "/usr/local/Cellar/python@3.8/3.8.2/Frameworks/Python.framework/Versions/3.8/lib/python3.8/configparser.py", line 395, in before_get
    self._interpolate_some(parser, option, L, value, section, defaults, 1)
  File "/usr/local/Cellar/python@3.8/3.8.2/Frameworks/Python.framework/Versions/3.8/lib/python3.8/configparser.py", line 412, in _interpolate_some
    p = rest.find("%")
AttributeError: 'list' object has no attribute 'find'
```

```toml
[tool.poetry]
name = "fake"
version = "0.1.0"
description = "dummy"
authors = ["Bruce Adams <bruce.adams@acm.org>"]

[tool.poetry.dependencies]
python = "^3.8"
connexion = { extras = ["swagger-ui"], version = "^2.7.0" }
okta_jwt = "^1.3.5"

[tool.poetry.dev-dependencies]
black = "^19.10b0"
fire = "^0.3.1"
mypy = "^0.770"
pyjwt = { extras = ["crypto"], version = "^1.7.1" }
pylint = "^2.5.0"
pytest = "^5.4.1"
pytest-cov = "^2.8.1"

[tool.pylint."MESSAGES CONTROL"]
# bad-continuation - pylint sometimes complains about black's formatting; we go with black
# logging-format-interpolation, logging-fstring-interpolation, logging-not-lazy
# - pylint wants deferred formatting for log messages (to use less CPU if that
#   log level is disabled). We prefer the legibility of f-strings: f"a is {a}"
disable = [
    "bad-continuation",
    "logging-format-interpolation",
    "logging-fstring-interpolation",
    "logging-not-lazy"
]

[build-system]
requires = ["poetry>=1.0"]
build-backend = "poetry.masonry.api"
```

### Expected behavior
What I hope for is that the `pyproject.toml` above behaves the same as when the disable line looks like this.

```
disable = "bad-continuation,logging-format-interpolation,logging-fstring-interpolation,logging-not-lazy"
```

### pylint --version output

```bash
$ pylint --version
pylint 2.5.0
astroid 2.4.0
Python 3.8.2 (default, Mar 11 2020, 00:29:50)
[Clang 11.0.0 (clang-1100.0.33.17)]
```

### Additional information

I tried adding a test for the behavior I want into https://github.com/PyCQA/pylint/blob/master/tests/test_config.py I was _very_ surprised to have that test pass!

I can add that test in a PR, if you'd like. Here it is, just for reference:

```python
import unittest.mock

import pylint.lint


def test_can_read_toml(tmp_path):
    config_file = tmp_path / "pyproject.toml"
    config_file.write_text(
        "[tool.pylint.'messages control']\n"
        "disable='logging-format-interpolation,logging-fstring-interpolation'\n"
        "enable='missing-module-docstring'\n"
        "jobs=10\n"
    )
    linter = pylint.lint.PyLinter()
    linter.global_set_option = unittest.mock.MagicMock()
    linter.read_config_file(str(config_file))

    assert linter.global_set_option.called_with(
        "disable", "logging-format-interpolation"
    )
    assert linter.global_set_option.called_with(
        "disable", "logging-fstring-interpolation"
    )
    assert linter.global_set_option.called_with("enable", "missing-module-docstring")
    assert linter.global_set_option.called_with("jobs", 10)


def test_can_read_toml_array(tmp_path):
    config_file = tmp_path / "pyproject.toml"
    config_file.write_text(
        '[tool.pylint."messages control"]\n'
        'disable=[\n"logging-format-interpolation",\n"logging-fstring-interpolation"\n]\n'
        'enable="missing-module-docstring"\n'
        "jobs=10\n"
    )
    linter = pylint.lint.PyLinter()
    linter.global_set_option = unittest.mock.MagicMock()
    linter.read_config_file(str(config_file))

    assert linter.global_set_option.called_with(
        "disable", "logging-format-interpolation"
    )
    assert linter.global_set_option.called_with(
        "disable", "logging-fstring-interpolation"
    )
    assert linter.global_set_option.called_with("enable", "missing-module-docstring")
    assert linter.global_set_option.called_with("jobs", 10)


def test_can_read_setup_cfg(tmp_path):
    config_file = tmp_path / "setup.cfg"
    config_file.write_text(
        "[pylint.messages control]\n"
        "disable=all\n"
        "enable=missing-module-docstring\n"
        "jobs=10\n"
    )
    linter = pylint.lint.PyLinter()
    linter.global_set_option = unittest.mock.MagicMock()
    linter.read_config_file(str(config_file))

    assert linter.global_set_option.called_with("disable", "all")
    assert linter.global_set_option.called_with("enable", "missing-module-docstring")
    assert linter.global_set_option.called_with("jobs", 10)
```

### PR DESCRIPTION
**Pull request by dbaty**
## Description

Before that, we had to use strings in a TOML configuration file, like
this:

```toml
enable = "use-symbolic-message-instead,useless-suppression"
jobs = "10"
suggestion-mode = "no"
```

TOML supports rich types like list, integer and boolean. They make for
a more readable and less error-prone file. We can now express the same
configuration like this:

```toml
enable = [
    "use-symbolic-message-instead",
    "useless-suppression",
]
jobs = 10
suggestion-mode = false
```

## Type of Changes
|   | Type |
| ------------- | ------------- |
| ✓  | :bug: Bug fix  |

## Related Issue

Closes #3538 

## Tests

Previous tests were not quite right because they used `MagicMock`, which happily says yes to, well, whatever you throw at it, including `magic_mocked_thing.called_with("anything", "really")`. ;)

## Alternative

I aimed for the least invasive change, but I think that we may be trying too hard to keep using ConfigParser. Here we read TOML files (with rich types such as list), then convert back to strings to shoehorn the values into ConfigParser, then we deserialize again to rich types for final use. I feel like we may want to get rid of ConfigParser. Is this a goal that we would want to pursue?

Cons:
- a (limited?) risk of getting it wrong, especially for INI files which are working (aka: breaking something that has been working well for a long time, I guess).
- ?

Pros:
- a saner way to read files;
- we'll be ready when ``.pysetup.yaml`` becomes the next Standard Way (tm) to hold configuration in a Python project :trollface: 

I did not look at all at how it could look or if it's even technically feasible (if one was willing to do it).

### CODE REVIEW

File: tests/test_config.py [hunk 1]
```diff
@@ -1,37 +1,98 @@
...
+    with unittest.mock.patch("sys.exit") as mocked_exit:
+        # Do not actually run checks, that could be slow.
+        with unittest.mock.patch("pylint.lint.pylinter.PyLinter.check"):
+            runner = pylint.lint.Run(args)
+
+    enabled = {
+        msgid for msgid, _enabled in runner.linter._msgs_state.items() if _enabled
+    }
+    # logging-format-interpolation, logging-fstring-interpolation
+    assert enabled == {"W1202", "W1203"}
```
Comment (PCManticore) [inline tests/test_config.py:23]: Rather than inspecting `_msgs_state` which should be an internal concern of the linter, you could use `runner.linter.is_message_enabled()` instead.

File: tests/test_config.py [hunk 2]
```diff
@@ -1,37 +1,98 @@
...
+    assert runner.linter.config.jobs == 10
+    type_checker = runner.linter._checkers["typecheck"][0]
```
Comment (PCManticore) [inline tests/test_config.py:25]: I would prefer if we'd not check this as it depends on a private variable set in a checker. The number of jobs though should be fine.

**Comment by dbaty** [comment]
Thanks for the quick review, @PCManticore . :)

tl;dr: I applied your recommendations, tweaked a few minor things, rebased off of master and squashed my commits. All builds are green.

Details:

1. I applied your recommendations: the tests now use `is_message_enabled()` and check for another boolean option (`reports`) that is accessible through a public API.

2. The tests now use only "disable" instead of both "disable" and "enable", which I could not make work. (First I tried keeping "enable" and setting `disable = "all"`, but then the tests would randomly fails, at least on Python 3.5, because sometimes "disable" was processed before "enable", and sometimes it was the other way around, so `_msgs_state` was populated differently.)

3. I moved the changelog entry to the new 2.5.3 section. TOML support is new in 2.5 and I think that such a bug fix belongs to a patch release. I can move it back to the 2.6 section if you disagree.
